logical/logical.go

```diff
@@ -2,6 +2,32 @@ package logical
 
 import log "github.com/mgutz/logxi/v1"
 
+// BackendType is the type of backend that is being implemented
+type BackendType uint32
+
+// The these are the types of backends that can be derived from
+// logical.Backend
+const (
+	TypeUnknown     BackendType = 0 // This is also the zero-value for BackendType
+	TypeLogical     BackendType = 1
+	TypeCredential  BackendType = 2
+	TypePassthrough BackendType = 3
```

Comment:
I think I had a more detailed explanation of this but it got removed by Jeff. Basically, this replaces [the need to perform type assertion](https://github.com/hashicorp/vault/pull/2874/files#diff-90984b74207af54c7b459b3c29bb3fefR259) during `HandleRequest`, and it also [simplifies](https://github.com/hashicorp/vault/pull/2874/files#diff-001b4cc91af2ffd8866a8d0463f12a87R17) the `PassthroughBackend` factories in `logical_passthrough.go`